geps/gep-3792/index.md

```diff
@@ -246,16 +233,254 @@ guarantee that **all four** of the problems listed above need resolutions, and
+implementations and the mesh implementations, so it's worth looking at some
+alternatives.
+
+- The OCG could be required to exactly mimic the ALPN/transport metadata
+  combination used by an existing mesh. However, the existing meshes don't
+  share a single common mechanism, so this would require a lot of work on the
+  OCG's part, and it wouldn't be portable between meshes.
+
+- The OCG could simply skip ALPN, and hand a "bare" mTLS connection to the
+  mesh. In general, existing meshes don't support this in the way that we
+  want: depending on the destination of the connection, they may interpret it
+  as application-level mTLS that they should treat as an opaque data stream,
+  or they may simply refuse it. This alternative, therefore, both shifts the
+  entire burden of implementation to the meshes, and probably makes it
+  impossible to correctly handle application-level mTLS.
+
+  (Whether meshes _should_ support application-level mTLS in this way is a
+  separate discussion, and is out of scope for this GEP.)
+
+- We could perhaps abuse the [PROXY protocol] for this, or define something
+  similar. This would appear to increase the implementation burden on both
+  sides without providing appreciable benefit.
+
+[PROXY protocol]: https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt
 
-### Gateway For Mesh (East/West)
+None of these alternatives would appear to be better than using ALPN in a
+similar way to how (some) existing meshes already use it.
+
+### Solving the Discovery Problem
+
+The discovery problem is that not every workload in the cluster is required to
+be meshed, which means that the OCG needs a way to know whether a given
+connection must participate in the mesh or not. In practice, this isn't
+actually a question of _workloads_ but of _Routes_: the point of interface
+between a Gateway and a workload in the cluster is not a Pod or a Service, but
+rather a Route.
+
+We could approach this in a few different ways:
+
+1. Assume that _all_ Routes in the cluster are meshed. This is obviously a
+   nonstarter.
+
+2. Assume that _all_ Routes bound to the OCG are meshed. This is also a
+   nonstarter.
+
+3. Add a field to the Route resource that indicates whether the Route is
+   meshed. This feels like quite a bit of an imposition on [Ana] (and, again,
+   if we do our jobs correctly, [Ana] shouldn't need to think about this).
+
+4. Add a field to the Gateway resource that enumerates Routes that are meshed.
+   This would be a lot of work for [Chihiro] and [Ian] to maintain; while
+   that's better than putting the burden on [Ana], it still isn't good.
+
+5. Add a field to the Gateway resource that enumerates namespaces that are
+   meshed. This isn't quite as bad as option 4, but it's still a lot of work
+   for [Chihiro] and [Ian], _and_ it limits us to only having entire
+   namespaces meshed.
+
+6. Add a label selector to the Gateway resource and declare that Routes with
```

Comment:
One other option that randomly popped in my head (not saying it's a great one): add some metadata to backendRef. This shouldn't at all take any of backendTLSPolicy's scope, but it's a way for the route to say "this backend is meshed"

Reply:
I would _really_ like to arrange things so that Ana needn't take on that burden. But let's keep it in mind.